static/libs-widget-ext.ts

```diff
@@ -0,0 +1,556 @@
+// Copyright (c) 2021, Compiler Explorer Authors
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+//     * Redistributions of source code must retain the above copyright notice,
+//       this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above copyright
+//       notice, this list of conditions and the following disclaimer in the
+//       documentation and/or other materials provided with the distribution.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+import { options } from './options';
+import * as local from './local';
+import { Library, LibraryVersion } from './options.interfaces';
+
+
+const FAV_LIBS_STORE_KEY = 'favlibs';
+
+interface StateLib {
+    id?: string;
+    name?: string;
+    ver?: string;
+    version?: string;
+}
+
+interface WidgetState {
+    libs?: StateLib[];
+}
+
+export type CompilerLibs = Record<string, Library>;
+type LangLibs = Record<string, CompilerLibs>;
+type AvailableLibs = Record<string, LangLibs>;
+type LibInUse = {libId: string, versionId: string} & LibraryVersion;
+
+type FavLibraries = Record<string, string[]>;
+
+export class Widget {
```

Comment:
Maybe we could rename this to LibrariesWidget? Widget sounds very generic

Reply:
You're right